src/stackmap_checker/call_stack_state.c

```diff
@@ -115,49 +129,202 @@ void insert_frames(call_stack_state_t *state, size_t index,
+                // frame they were inlined in.
+                memcpy(res_state->frames[j].registers,
+                       state_copy->frames[i + 1].registers,
+                       REGISTER_COUNT * sizeof(unw_word_t));
+            }
+            // Insert the restored frames inside `state`.
+            insert_frames(state, i + 1, res_state->frames, res_state->depth);
+            free_call_stack_state(res_state);
+        }
+    }
+    free(state_copy);
+    return inlined;
+}
+
+call_stack_state_t* get_restored_state(stack_map_t *sm,
+                                       uint64_t start_addr,
+                                       uint64_t ppid)
+{
+    call_stack_state_t *state = malloc(sizeof(call_stack_state_t));
+    state->depth = 0;
+    state->frames = NULL;
+    stack_map_record_t *rec = NULL;
+    // The stack map records which correspond to each call on the stack.
+    // The number of frames on the stack.
+    // Find the 'return addresses' of the inlined functions
+    // The next 'real' record in this function.
+    uint64_t opt_ret_addr = start_addr;
+    do {
+        rec = stmap_first_rec_after_addr(sm, opt_ret_addr);
+        if (!rec) {
+            break;
+        }
+        stack_size_record_t *size_rec = stmap_get_size_record(sm, rec->index);
+        if (!size_rec) {
+            errx(1, "Size record not found\n");
+        }
+        stack_map_record_t *unopt_rec =
+            stmap_get_map_record(sm, ~rec->patchpoint_id);
+        stack_size_record_t *unopt_size_rec =
+            stmap_get_size_record(sm, unopt_rec->index);
+        ++state->depth;
+        state->frames = realloc(state->frames, state->depth * sizeof(frame_t));
+        // ret_addr does not hold the address of the return address
+        // if inlined = 1 (it stores the return address instead).
+        state->frames[state->depth - 1].ret_addr =
+            unopt_rec->instr_offset + unopt_size_rec->fun_addr + 13;
+        opt_ret_addr = rec->instr_offset + size_rec->fun_addr + 1;
+        state->frames[state->depth - 1].size = unopt_size_rec->stack_size;
+        state->frames[state->depth - 1].record = *rec;
+        state->frames[state->depth - 1].real_record =
+            *stmap_get_map_record_after_addr(sm, rec->patchpoint_id,
+                                             start_addr);
+        state->frames[state->depth - 1].inlined = 1;
+    } while(rec->patchpoint_id != ppid);
+    return state;
+}
+
+void insert_real_addresses(call_stack_state_t *state, restored_segment_t seg)
+{
+    // Now add the return addresses and saved base pointers to the new 'stack'.
+    // Start with the low addresses: `cur_bp` is the BP of the last function
+    // callled
+    char *cur_bp = (char *)seg.start_addr + state->frames[0].size - 8;
+    for (size_t i = 0; i + 1 < state->depth; ++i) {
+        state->frames[i].bp = (uint64_t)cur_bp;
+        if (state->frames[i + 1].inlined) {
```

Comment:
Is it me or are the arms of this conditional the same, just one has a cast?

Reply:
In the else branch, the return address cast to `(uint64_t *)` is also dereferenced. This is because the `else` branch deals with real frames. The *address* of the return address of each real frame is known (rbp + 8), and it is used to obtain the value of the return address. On the other hand, if a frame is  associated with an inlined function (and created by the recovery mechanism), the *address* of its return address is not known yet, as the frame has not yet been placed on the stack. Its return address is known, because stack map records encode the return addresses of functions. It isn't ideal, though

Reply:
ok